apps/cyberstorm-nextjs/middleware.ts

```diff
@@ -0,0 +1,54 @@
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+import {
+  decode,
+  handleJWTGenerationAndReturnResponseOrRedirect,
+  redirectToPremium,
+} from "./utils/auth";
+
+export const config = {
+  matcher: [
+    /*
+     * Match all request paths except for the ones starting with:
+     * - _next/static (static files)
+     * - _next/image (image optimization files)
+     * - favicon.ico (favicon file)
+     */
+    "/((?!_next/static|_next/image|favicon.ico|images).*)",
+  ],
+};
+
+export async function middleware(request: NextRequest) {
+  const req = request;
+  const response = NextResponse.next();
+  // Do this temporarily while in dev, since the django session might not be in the browser
+  // request.cookies.set("sessionid", "");
+
+  if (process.env.CYBERSTORM_AUTH_ENABLED) {
+    // JWT exists?
+    if (request.cookies.has("CyberstormAuth")) {
+      // Validate JWT
+      try {
+        const JWTToken = request.cookies.get("CyberstormAuth");
+        if (JWTToken) {
+          await decode(JWTToken.value);
+          return response;
+        } else {
+          // Token value was empty, redirect to "Go Premium"
+          return redirectToPremium(request, response);
+        }
+      } catch (e: any) {
+        if (e.code === "ERR_JWT_EXPIRED") {
+          // Expired so figure and validate django session and generate a new JWT token
+          return handleJWTGenerationAndReturnResponseOrRedirect(req, response);
+        }
+        throw e;
```

Comment:
Hmm, I'm not sure actually. The dev env throws the error of course and displays it, but no clue if it displays the customized "error 500" page we have. I'll ask around.

Reply:
400 is "bad request", so did you mean 403 or did you mean "forbidden"?

Generally speaking I'd say that it depends on the situations that actually cause `decode()` to throw. For "hey, this token isn't issued by us", 403 would probably be the right choice. If the token is unreadable/malformed, then it's bad request. And for "hey, the server doesn't have the private key set up", 500 would be best.

If it's impossible to to tell the cases apart, i.e. there's no more stuff like `errors.JWTExpired`, then perhaps go with 403. Otherwise I would say 500 (since it's basically us saying we don't know what went wrong), but 403 gives a hint that the problem is somehow related to the authentication, and may user/us help resolve the issue.

Reply:
Wise words, I've added few more error checks.

Reply:
Also I might revisit this, once I get the Django side done.